uaclient/entitlements/livepatch.py

```diff
@@ -59,8 +59,12 @@ class LivepatchEntitlement(base.UAEntitlement):
     description = (
         'Canonical Livepatch Service'
         ' (https://www.ubuntu.com/server/livepatch)')
-    static_affordances = (
-        ('Cannot install Livepatch on a container', util.is_container, False),)
+
+    def __init__(self, cfg=None):
+        super().__init__(cfg)
+        self.static_affordances = (
```

Comment:
Why the move?

Reply:
Moved this into init so that we could mock out util.is_container at init timeframe in unit tests. If it remains a predefined class attr, mocking doesn't help because cls.static_affordances is already initialized to the real util.is_container.

Reply:
Adding a comment to that affect in the code

Reply:
I don't love the fact that `static_affordances` is no longer, well, static.

If we instead change the function in `static_affordances` to a `lambda` which calls `util.is_container` then we can successfully mock (because the `util.is_container` name lookup now happens at call time and not at class definition time):

```diff
diff --git a/uaclient/entitlements/livepatch.py b/uaclient/entitlements/livepatch.py
index b8d86ab..23b2e16 100644
--- a/uaclient/entitlements/livepatch.py
+++ b/uaclient/entitlements/livepatch.py
@@ -58,13 +58,12 @@ class LivepatchEntitlement(base.UAEntitlement):
     description = (
         'Canonical Livepatch Service'
         ' (https://www.ubuntu.com/server/livepatch)')
-
-    def __init__(self, cfg=None):
-        super().__init__(cfg)
-        # Set static_affordances in init so we can mock is_container in tests
-        self.static_affordances = (
-            ('Cannot install Livepatch on a container', util.is_container,
-             False),)
+    static_affordances = (
+        ('Cannot install Livepatch on a container',
+         # Use a lambda so we can mock util.is_container in tests
+         lambda: util.is_container(),
+         False),
+    )
 
     def enable(self):
         """Enable specific entitlement.
```

Reply:
Thanks for the Dan, just pushed a change and tested livepatch enable on container and non-container and success.